File: JC/src/se/liu/albhe576/project/frontend/Scanner.java
package se.liu.albhe576.project.frontend;

import java.util.Map;
import se.liu.albhe576.project.backend.Compiler;

/**
 * A Scanner reads through the given input stream and creates Tokens and supplies them to the parser.
 *
 * @see Token
 * @see TokenType
 *
 */
public class Scanner {
    private final String input;
    private int index;
    private int line;
    private final String filename;

    public int getLine(){
        return this.line;
    }
    private char getCurrentChar(){
        if(index >= input.length()){
            return 0;
        }
        return input.charAt(index);
    }

    public void skipWhiteSpace(){
        // skip until we hit something we care about
	while (true) {
	    char currentChar = this.getCurrentChar();
	    if(currentChar == 0 || Character.isDigit(currentChar) || Character.isAlphabetic(currentChar)){
		return;
	    }
	    switch(currentChar){
		case '/' -> {
		    advance();
		    if(this.matchNext('/')){
			while(!this.isOutOfBounds() && this.getCurrentChar() != '\n'){
			    advance();
			}
			line++;
			break;
		    }
		    this.index--;
		    return;
		}
		case '\"', '{', '[',
		']', '}', '*', ';',
		'<', '>', '(', ')',
		'!', '%', '.', '#',
		'^', ',', '&', '|',
		'-', '+', '=' ->{
		    return;
		}
		case '\n'->{
		    this.line++;
		}
		default -> {
		    // Just continue skipping
		}
	    }
	    index++;
	}
    }

    private Token parseString() {
        int startIndex  = this.index;
        int line        = this.line;

        // parse until we hit the next '"'
        char currentChar = advance();
        while(!isOutOfBounds() && currentChar != '\"'){
            currentChar = advance();
        }

        // check for unterminated strings
        if(isOutOfBounds()){
            Compiler.panic("Unterminated string at starting", line, filename);
        }

        String literal = this.input.substring(startIndex, this.index - 1);
        return createToken(TokenType.TOKEN_STRING_LITERAL, literal);
    }
    private final static Map<String, TokenType> RESERVED_KEYWORDS = Map.ofEntries(
            Map.entry("struct", TokenType.TOKEN_STRUCT),
            Map.entry("string", TokenType.TOKEN_STRING),
            Map.entry("if", TokenType.TOKEN_IF),
            Map.entry("else", TokenType.TOKEN_ELSE),
            Map.entry("for", TokenType.TOKEN_FOR),
            Map.entry("while", TokenType.TOKEN_WHILE),
            Map.entry("return", TokenType.TOKEN_RETURN),
            Map.entry("short", TokenType.TOKEN_SHORT),
            Map.entry("int", TokenType.TOKEN_INT),
            Map.entry("long", TokenType.TOKEN_LONG),
            Map.entry("double", TokenType.TOKEN_DOUBLE),
            Map.entry("byte", TokenType.TOKEN_BYTE),
            Map.entry("float", TokenType.TOKEN_FLOAT),
            Map.entry("void", TokenType.TOKEN_VOID)
    );
    public TokenType getKeyword(String literal){
        if(RESERVED_KEYWORDS.containsKey(literal)){
            return RESERVED_KEYWORDS.get(literal);
        }
        return TokenType.TOKEN_IDENTIFIER;
    }

    private Token parseKeyword() {
        int startIndex = this.index - 1;
        char currentChar = advance();
        // keep parsing while we're not out of boudns and we have a digit, character or _
        while(!isOutOfBounds() && (Character.isDigit(currentChar)|| Character.isAlphabetic(currentChar) || currentChar == '_')){
            currentChar = advance();
        }

        this.index--;
        String literal = this.input.substring(startIndex, this.index);
        TokenType keyword = getKeyword(literal);
        return createToken(keyword, literal);
    }

    private boolean isValidHexChar(char c){
        return ('a' <= c &&  c <= 'f') || ('A' <= c && c <= 'F');
    }

    private Token parseHex(int startIndex){
        char currentChar = advance();
        while(!isOutOfBounds() && (Character.isDigit(currentChar) || (this.isValidHexChar(currentChar)))){
            currentChar = advance();
        }
        this.index--;
        String literal = this.input.substring(startIndex, this.index);
        return createToken(TokenType.TOKEN_INT_LITERAL, literal);

    }
    private Token parseBinary(int startIndex){
        char currentChar = advance();

        while(!isOutOfBounds() && (currentChar == '0' || currentChar == '1')){
            currentChar = advance();
        }
        this.index--;
        String literal = this.input.substring(startIndex, this.index);
        return createToken(TokenType.TOKEN_INT_LITERAL, literal);
    }

    private boolean continueParsingNumber(char current){
        return !isOutOfBounds() && Character.isDigit(current);
    }
    private Token parseNumber(){
        int startIndex = this.index - 1;
        TokenType type = TokenType.TOKEN_INT_LITERAL;

        char currentChar = advance();
        // check for binary and hex values
        if(this.input.charAt(startIndex) == '0'){
            if(currentChar == 'x'){
                return this.parseHex(startIndex);
            }
            else if(currentChar == 'b'){
                return this.parseBinary(startIndex);
            }
        }

        while(continueParsingNumber(currentChar)){
            currentChar = advance();
        }

        // check for float
        if(currentChar == '.'){
            currentChar = getCurrentChar();
            type = TokenType.TOKEN_FLOAT_LITERAL;
            while(continueParsingNumber(currentChar)){
                currentChar = advance();
            }
        }
        this.index--;
        String literal = this.input.substring(startIndex, this.index);
        return createToken(type, literal);

    }

    private Token createToken(TokenType tokenType, String literal){
        return new Token(tokenType, this.line, literal);
    }

    private boolean matchNext(char toMatch){
        if(getCurrentChar() == toMatch){
            this.index++;
            return true;
        }
        return false;
    }

    private char advance(){
        char out = getCurrentChar();
        this.index++;
        return out;
    }

    private boolean isOutOfBounds(){
        return this.index >= this.input.length();
    }

    public Token parseToken() {
        skipWhiteSpace();

        if(this.isOutOfBounds()){
            return this.createToken(TokenType.TOKEN_EOF, "EOF");
        }

        char currentChar = advance();
        // check number and identifier
        if(Character.isAlphabetic(currentChar)){
            return parseKeyword();
        }
        if(Character.isDigit(currentChar)){
            return parseNumber();
        }

        return switch (currentChar) {
            // check single character tokens
            case '{' -> this.createToken(TokenType.TOKEN_LEFT_BRACE, "{");
            case '}' -> this.createToken(TokenType.TOKEN_RIGHT_BRACE, "}");
            case '[' -> this.createToken(TokenType.TOKEN_LEFT_BRACKET, "[");
            case ']' -> this.createToken(TokenType.TOKEN_RIGHT_BRACKET, "]");
            case ';' -> this.createToken(TokenType.TOKEN_SEMICOLON, ";");
            case '%' -> this.createToken(TokenType.TOKEN_MOD, "%");
            case ',' -> this.createToken(TokenType.TOKEN_COMMA, ",");
            case '(' -> this.createToken(TokenType.TOKEN_LEFT_PAREN, "(");
            case ')' -> this.createToken(TokenType.TOKEN_RIGHT_PAREN, ")");
            case '\"' -> parseString();
            // parse pound defines
            case '#' ->
            {
                if(this.input.startsWith("include", this.index)){
                    this.index += "include".length();
                    yield this.createToken(TokenType.TOKEN_INCLUDE, "#include");
                }
                else if(this.input.startsWith("extern", this.index)){
                    this.index += "extern".length();
                    yield this.createToken(TokenType.TOKEN_EXTERN, "#extern");
                }
                else if(this.input.startsWith("define", this.index)){
                    this.index += "define".length();
                    yield this.createToken(TokenType.TOKEN_DEFINE, "#define");
                }
                Compiler.panic("tried to parse '#' but didnt find include/exctern/define afterwards\n", line, filename);
                // unreachable
                yield null;
            }
            // check two character tokens
            case '/' ->{
                if(matchNext('=')){
                    yield this.createToken(TokenType.TOKEN_AUGMENTED_SLASH, "/=");
                }
                yield this.createToken(TokenType.TOKEN_SLASH, "/");
            }
            case '*' -> {
                if(matchNext('=')){
                    yield this.createToken(TokenType.TOKEN_AUGMENTED_STAR, "*=");
                }
                yield this.createToken(TokenType.TOKEN_STAR, "*");
            }
            case '.' ->{
                if(this.input.startsWith("..", this.index)){
                    this.index += "..".length();
                    yield this.createToken(TokenType.TOKEN_ELLIPSIS, "...");
                }
                yield this.createToken(TokenType.TOKEN_DOT, ".");
            }
            case '^' -> {
                if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_AUGMENTED_XOR, "^=");
                }
                yield this.createToken(TokenType.TOKEN_XOR, "^");
            }
            case '&' -> {
                if (matchNext('&')) {
                    yield this.createToken(TokenType.TOKEN_AND_LOGICAL, "&&");
                }
                else if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_AUGMENTED_AND, "&=");
                }
                yield this.createToken(TokenType.TOKEN_AND_BIT, "&");
            }
            case '|' -> {
                if (matchNext('|')) {
                    yield this.createToken(TokenType.TOKEN_OR_LOGICAL, "||");
                }
                else if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_AUGMENTED_OR, "|=");
                }
                yield this.createToken(TokenType.TOKEN_OR_BIT, "|");
            }
            case '!' -> {
                if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_BANG_EQUAL, "!=");
                }
                yield this.createToken(TokenType.TOKEN_BANG, "!");
            }
            case '>' -> {
                if (matchNext('>')) {
                    yield this.createToken(TokenType.TOKEN_SHIFT_RIGHT, ">>");
                }
                else if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_GREATER_EQUAL, ">=");
                }
                yield this.createToken(TokenType.TOKEN_GREATER, ">");
            }
            case '<' -> {
                if (matchNext('<')) {
                    yield this.createToken(TokenType.TOKEN_SHIFT_LEFT, "<<");
                }
                else if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_LESS_EQUAL, "<=");
                }
                yield this.createToken(TokenType.TOKEN_LESS, "<");
            }
            case '-' -> {
                if (matchNext('-')) {
                    yield this.createToken(TokenType.TOKEN_DECREMENT, "--");
                } else if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_AUGMENTED_MINUS, "+=");
                }else if(Character.isDigit(getCurrentChar())){
                   yield this.parseNumber();
                }
                yield this.createToken(TokenType.TOKEN_MINUS, "-");
            }
            case '+' -> {
                if (matchNext('+')) {
                    yield this.createToken(TokenType.TOKEN_INCREMENT, "++");
                } else if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_AUGMENTED_PLUS, "+=");
                }
                yield this.createToken(TokenType.TOKEN_PLUS, "+");
            }
            case '=' -> {
                if (matchNext('=')) {
                    yield this.createToken(TokenType.TOKEN_EQUAL_EQUAL, "==");
                }
                yield this.createToken(TokenType.TOKEN_EQUAL, "=");
            }
            default ->{
                Compiler.panic(String.format("Illegal character '%c'", currentChar), line, filename);
                // Unreachable
                yield null;
            }
        };
    }
    public Scanner(String input, String filename){
        this.filename = filename;
        this.input = input;
        this.index = 0;
        this.line = 1;
    }
}

File: CC/src/scanner.c

#include "scanner.h"
#include "common.h"
#include "stdbool.h"
#include "token.h"
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void error(const char* message)
{
  printf("%s\n", message);
  exit(1);
}

void init_scanner(Scanner* scanner, Arena* arena, String* literal, const char* filename)
{
  scanner->filename = (u8*)filename;
  scanner->index    = 0;
  scanner->line     = 1;
  scanner->input    = literal;
  scanner->arena    = arena;
}

static bool match_substring(String* literal, int index, const char* to_match, int length)
{
  if (literal->len <= index + length)
  {
    return false;
  }
  return strncmp(&literal->buffer[index], to_match, length) == 0;
}

static bool is_out_of_bounds(Scanner* scanner)
{
  return scanner->input->len <= scanner->index;
}

static u8 current_char(Scanner* scanner)
{
  return scanner->input->buffer[scanner->index];
}

static u8 advance(Scanner* scanner)
{
  char out = current_char(scanner);
  scanner->index++;
  return out;
}

static bool match_next(Scanner* scanner, char toMatch)
{
  if (current_char(scanner) == toMatch)
  {
    scanner->index++;
    return true;
  };
  return false;
}

static void skip_whitespace(Scanner* scanner)
{

  while (true)
  {
    if (is_out_of_bounds(scanner))
    {
      return;
    }
    u8 current = current_char(scanner);
    if (isdigit(current) || isalpha(current))
    {
      return;
    }
    switch (current)
    {
    case '/':
    {
      advance(scanner);
      if (match_next(scanner, '/'))
      {
        while (!is_out_of_bounds(scanner) && current_char(scanner) != '\n')
        {
          advance(scanner);
        }
        scanner->line++;
        break;
      }
      scanner->index--;
      return;
    }

    case '\"':
    case '{':
    case '[':
    case ']':
    case '}':
    case '*':
    case ';':
    case '<':
    case '>':
    case '(':
    case ')':
    case '!':
    case '%':
    case '.':
    case '#':
    case '^':
    case ':':
    case ',':
    case '&':
    case '|':
    case '-':
    case '+':
    case '=':
    {
      return;
    }
    case '\n':
    {
      scanner->line++;
    }
    default:
    {
      scanner->index++;
    }
    }
  }
}

static Token* parse_string(Scanner* scanner)
{
  i32 index   = scanner->index;
  u8  current = advance(scanner);

  while (!is_out_of_bounds(scanner) && current != '\"')
  {
    current = advance(scanner);
  }
  if (is_out_of_bounds(scanner))
  {
    error("Unterminated string!");
  }
  String literal = {};
  literal.buffer = (char*)&scanner->input->buffer[index];
  literal.len    = scanner->index - index - 1;

  return create_token(scanner->arena, TOKEN_STRING_LITERAL, literal, scanner->line);
}

static TokenType get_keyword(String literal)
{
  static const char* keywords[]         = {"struct", "string", "if", "else", "for", "while", "return", "short", "int", "long", "double", "float", "byte", "void"};
  static TokenType   tokens[]           = {TOKEN_STRUCT, TOKEN_STRING, TOKEN_IF,   TOKEN_ELSE,   TOKEN_FOR,   TOKEN_WHILE, TOKEN_RETURN,
                                           TOKEN_SHORT,  TOKEN_INT,    TOKEN_LONG, TOKEN_DOUBLE, TOKEN_FLOAT, TOKEN_BYTE,  TOKEN_VOID};
  u32                number_of_keywords = ArrayCount(keywords);
  for (u32 i = 0; i < number_of_keywords; i++)
  {
    if (literal.len == strlen(keywords[i]) && strncmp(keywords[i], literal.buffer, literal.len) == 0)
    {
      return tokens[i];
    }
  }
  return TOKEN_IDENTIFIER;
}

static Token* parse_keyword(Scanner* scanner)
{
  i32 index   = scanner->index - 1;
  u8  current = advance(scanner);
  while (!is_out_of_bounds(scanner) && (isdigit(current) || isalpha(current) || current == '_'))
  {
    current = advance(scanner);
  }

  String literal = {};
  literal.buffer = (char*)&scanner->input->buffer[index];
  scanner->index--;
  literal.len    = scanner->index - index;
  return create_token(scanner->arena, get_keyword(literal), literal, scanner->line);
}

static bool is_valid_hex(char c)
{
  return ('a' <= c && c <= 'f') || ('A' <= c && c <= 'F');
}

static Token* parse_hex(Scanner* scanner, i32 index)
{
  char current = advance(scanner);
  while (!is_out_of_bounds(scanner) && (isdigit(current) || is_valid_hex(current)))
  {
    current = advance(scanner);
  }
  scanner->index--;
  String literal = {};
  literal.buffer = (char*)&scanner->input->buffer[index];
  literal.len    = scanner->index - index;
  return create_token(scanner->arena, TOKEN_INT_LITERAL, literal, scanner->line);
}

static Token* parse_binary(Scanner* scanner, i32 index)
{
  char current = advance(scanner);
  while (!is_out_of_bounds(scanner) && (current == '0' || current == '1'))
  {
    current = advance(scanner);
  }
  scanner->index--;
  String literal = {};
  literal.buffer = (char*)&scanner->input->buffer[index];
  literal.len    = scanner->index - index;
  return create_token(scanner->arena, TOKEN_INT_LITERAL, literal, scanner->line);
}

static Token* parse_number(Scanner* scanner)
{
  i32       index   = scanner->index - 1;
  TokenType type    = TOKEN_INT_LITERAL;

  char      current = advance(scanner);

  if (scanner->input->buffer[index] == '0')
  {
    if (current == 'x')
    {
      return parse_hex(scanner, index);
    }
    else if (current == 'b')
    {
      return parse_binary(scanner, index);
    }
  }

  while (!is_out_of_bounds(scanner) && isdigit(current))
  {
    current = advance(scanner);
  }

  if (current == '.')
  {
    current = current_char(scanner);
    type    = TOKEN_FLOAT_LITERAL;

    while (!is_out_of_bounds(scanner) && isdigit(current))
    {
      current = advance(scanner);
    }
  }
  scanner->index--;
  String literal = {};
  literal.buffer = (char*)&scanner->input->buffer[index];
  literal.len    = scanner->index - index;
  return create_token(scanner->arena, type, literal, scanner->line);
}

Token* parse_token(Scanner* scanner)
{
  skip_whitespace(scanner);
  if (is_out_of_bounds(scanner))
  {
    String literal = {};
    literal.buffer = "EOF";
    literal.len    = 3;
    return create_token(scanner->arena, TOKEN_EOF, literal, scanner->line);
  }

  char current = advance(scanner);

  if (isalpha(current))
  {
    return parse_keyword(scanner);
  }

  if (isdigit(current))
  {
    return parse_number(scanner);
  }

  switch (current)
  {
  case '{':
  {
    String literal = {};
    literal.buffer = "{";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_LEFT_BRACE, literal, scanner->line);
  }
  case '}':
  {
    String literal = {};
    literal.buffer = "}";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_RIGHT_BRACE, literal, scanner->line);
  }
  case '[':
  {
    String literal = {};
    literal.buffer = "[";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_LEFT_BRACKET, literal, scanner->line);
  }
  case ']':
  {
    String literal = {};
    literal.buffer = "]";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_RIGHT_BRACKET, literal, scanner->line);
  }
  case ';':
  {
    String literal = {};
    literal.buffer = ";";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_SEMICOLON, literal, scanner->line);
  }
  case '%':
  {
    String literal = {};
    literal.buffer = "%";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_MOD, literal, scanner->line);
  }
  case ',':
  {
    String literal = {};
    literal.buffer = ",";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_COMMA, literal, scanner->line);
  }
  case '(':
  {
    String literal = {};
    literal.buffer = "(";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_LEFT_PAREN, literal, scanner->line);
  }
  case ')':
  {
    String literal = {};
    literal.buffer = ")";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_RIGHT_PAREN, literal, scanner->line);
  }
  case '/':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "/=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AUGMENTED_SLASH, literal, scanner->line);
    }
    literal.buffer = "/";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_SLASH, literal, scanner->line);
  }
  case '*':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "*=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AUGMENTED_STAR, literal, scanner->line);
    }
    literal.buffer = "*";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_STAR, literal, scanner->line);
  }
  case '&':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "&=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AUGMENTED_AND, literal, scanner->line);
    }
    else if (match_next(scanner, '&'))
    {
      literal.buffer = "&&";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AND_LOGICAL, literal, scanner->line);
    }

    literal.buffer = "&";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_AND_BIT, literal, scanner->line);
  }
  case '|':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "|=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AUGMENTED_OR, literal, scanner->line);
    }
    else if (match_next(scanner, '|'))
    {
      literal.buffer = "||";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_OR_LOGICAL, literal, scanner->line);
    }

    literal.buffer = "|";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_OR_BIT, literal, scanner->line);
  }
  case '!':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "!=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_BANG_EQUAL, literal, scanner->line);
    }
    literal.buffer = "!";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_BANG, literal, scanner->line);
  }
  case '>':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = ">=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_GREATER_EQUAL, literal, scanner->line);
    }
    else if (match_next(scanner, '>'))
    {
      literal.buffer = ">>";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_SHIFT_RIGHT, literal, scanner->line);
    }
    literal.buffer = ">";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_GREATER, literal, scanner->line);
  }
  case '-':
  {
    String literal = {};
    if (match_next(scanner, '-'))
    {
      literal.buffer = "--";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_DECREMENT, literal, scanner->line);
    }
    else if (match_next(scanner, '='))
    {
      literal.buffer = "-=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AUGMENTED_MINUS, literal, scanner->line);
    }
    literal.buffer = "-";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_MINUS, literal, scanner->line);
  }
  case '+':
  {
    String literal = {};
    if (match_next(scanner, '+'))
    {
      literal.buffer = "++";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_INCREMENT, literal, scanner->line);
    }
    else if (match_next(scanner, '='))
    {
      literal.buffer = "-=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AUGMENTED_PLUS, literal, scanner->line);
    }
    literal.buffer = "+";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_PLUS, literal, scanner->line);
  }
  case '=':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "==";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_EQUAL_EQUAL, literal, scanner->line);
    }
    literal.buffer = "=";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_EQUAL, literal, scanner->line);
  }
  case '<':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "<=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_LESS_EQUAL, literal, scanner->line);
    }
    else if (match_next(scanner, '<'))
    {
      literal.buffer = "<<";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_SHIFT_LEFT, literal, scanner->line);
    }
    literal.buffer = "<";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_LESS, literal, scanner->line);
  }
  case '^':
  {
    String literal = {};
    if (match_next(scanner, '='))
    {
      literal.buffer = "^=";
      literal.len    = 2;
      return create_token(scanner->arena, TOKEN_AUGMENTED_XOR, literal, scanner->line);
    }
    literal.buffer = "^";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_XOR, literal, scanner->line);
  }
  case '.':
  {
    String literal      = {};
    int    match_length = 2;
    if (match_substring(scanner->input, scanner->index, "..", match_length))
    {
      scanner->index += match_length;
      literal.buffer = "...";
      literal.len    = 3;
      return create_token(scanner->arena, TOKEN_ELLIPSIS, literal, scanner->line);
    }
    literal.buffer = ".";
    literal.len    = 1;
    return create_token(scanner->arena, TOKEN_DOT, literal, scanner->line);
  }
  case '\"':
  {
    return parse_string(scanner);
  }
  case '#':
  {
    String literal = {};
    if (match_substring(scanner->input, scanner->index, "extern", 6))
    {
      scanner->index += 6;
      literal.buffer = "#extern";
      literal.len    = 7;
      return create_token(scanner->arena, TOKEN_EXTERN, literal, scanner->line);
    }
    else if (match_substring(scanner->input, scanner->index, "include", 7))
    {
      scanner->index += 7;
      literal.buffer = "#include";
      literal.len    = 8;
      return create_token(scanner->arena, TOKEN_INCLUDE, literal, scanner->line);
    }
    else if (match_substring(scanner->input, scanner->index, "define", 6))
    {
      scanner->index += 6;
      literal.buffer = "#define";
      literal.len    = 7;
      return create_token(scanner->arena, TOKEN_DEFINE, literal, scanner->line);
    }
  }
  default:
  {
    printf("%c %d\n", scanner->input->buffer[scanner->index], scanner->input->buffer[scanner->index]);
    error("Illegal character!");
  }
  }
  return 0;
}
